zephyr: allow recovery over usb cdc-acm with logs enabled

nvlsianpu opened the pull request:
Modyfied serial_adapter so log are alowed when using USB
CDC ACM serial port emulation.

As log needs to work asynchronously in its own thread,
mcuboot need to give some time for execution this thread using
k_sleep().

Added configuration for nrf52840_pca10056 which shows how
to enable looging along with USB - among other, thread log
processing is required.

build command (form zephyr-project root directory)
west build -d build/mcuboot/nrf52840_pca10056 -b nrf52840_pca10056
./bootloader/mcuboot/boot/zephyr/
 -- -DDTC_OVERLAY_FILE=./boards/nrf52840_pca10056_big.overlay
-DOVERLAY_CONFIG=./boards/nrf52840_pca10056_usb_cdc_log.conf

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
DNM as I still need to add either k-sleep() or k_yield() in few more places for improve log output, but this shows that log works along with usb cdc acm recovery.

@nordic-krch pleas look at what I have done.

nvlsianpu commented:
Notice that https://github.com/zephyrproject-rtos/zephyr/pull/22798 is required for robust USB CDC ACM  communication.

nvlsianpu commented on boot/boot_serial/src/boot_serial.c:
```diff
@@ -666,6 +666,9 @@ boot_serial_start(const struct boot_uart_funcs *f)
     off = 0;
     while (1) {
+#if defined(CONFIG_LOG) && !defined(CONFIG_LOG_IMMEDIATE)
+        k_sleep(3); /* give time to process the log */
```
should be zephyr only patch

mbolivar commented on boot/zephyr/boards/nrf52840_pca10056_big.overlay:
```diff
@@ -0,0 +1,35 @@
+/delete-node/ &scratch_partition;
+
+&flash0 {
+partitions {
```
Should be indented, right?

mbolivar commented on boot/zephyr/sample.yaml:
```diff
@@ -13,3 +13,8 @@ tests:
+    extra_args: OVERLAY_CONFIG=./boards/nrf52840_pca10056_usb_cdc_log.conf
+      DTC_OVERLAY=./boards/nrf52840_pca10056_big.overlay
+    platform_whitelist:  nrf52840_pca10056
+    tags: bootloader_mcuboot
```
Missing newline at end of file?

mbolivar commented on boot/zephyr/boards/nrf52840_pca10056_usb_cdc_log.conf:
```diff
@@ -0,0 +1,37 @@
+# The UART is used for Serial Recovery, so logging requires
```
Is there any way to do this conditionally with if statements in the Kconfig file instead of using a board-specific file? This makes it not portable to other boards.

nvlsianpu replied:
ok, will try to do that

nvlsianpu commented:
I Added dedicated thread for log processing of the highest application
priority. This allows to transmit all logs without adding k_sleep
anywhere else int the code.

Introduced boot log thread is simpler than the default log threat
which decreases flash footprint by a few dozen bytes.

@mbolivar what you think about such a solution.

nvlsianpu commented:
@hakonfam you opinion on this will be valuable.

hakonfam commented on boot/boot_serial/src/boot_serial.c:
```diff
@@ -666,6 +666,9 @@ boot_serial_start(const struct boot_uart_funcs *f)
     off = 0;
     while (1) {
+#if defined(CONFIG_LOG) && !defined(CONFIG_LOG_IMMEDIATE)
+        k_sleep(3); /* give time to process the log */
```
Magic value `3` should have a define with some comment on rationale. Otherwise it is extremely hard for future developers to understand the implications of it, and how to maintain it on other HW (if needed)

hakonfam replied:
Lol, I was doing per commit review, and I see now that this is indeed removed all together :)

hakonfam commented on boot/zephyr/boards/nrf52840_pca10056_big.overlay:
```diff
@@ -0,0 +1,35 @@
+			label = "image-1";
+			reg = <0x00075000 0x000065000>;
+		};
+		scratch_partition: partition@da000 {
```
Do we need this partition if we default to using the non-scratch-based but moved-based swap operation?

hakonfam commented on boot/zephyr/main.c:
```diff
@@ -175,6 +183,40 @@ static void do_boot(struct boot_rsp *rsp)
+#include <logging/log_ctrl.h>
+
+/* most simple log processing theread */
+void boot_log_therad_func(void *dummy1, void *dummy2, void *dummy3)
```
```suggestion
void boot_log_thread_func(void *dummy1, void *dummy2, void *dummy3)
```

hakonfam commented on boot/zephyr/main.c:
```diff
@@ -175,6 +183,40 @@ static void do_boot(struct boot_rsp *rsp)
+				K_HIGHEST_APPLICATION_THREAD_PRIO, 0, 30);
+		k_thread_name_set(&boot_log_thread, "logging");
+}
+#endif
```
```suggestion
#endif /* defined(CONFIG_LOG) && !defined(CONFIG_LOG_IMMEDIATE) &&\
    !defined(CONFIG_LOG_PROCESS_THREAD) */
```

hakonfam replied:
Not sure if this is really needed, but some comment on what the condition for the corresponding `if` would be welcome IMO

hakonfam commented on boot/zephyr/main.c:
```diff
@@ -175,6 +183,40 @@ static void do_boot(struct boot_rsp *rsp)
+     }
+}
+
+K_THREAD_STACK_DEFINE(boot_log_stack, 768);
```
Aren't stack sizes of threads typically user configurable? In this case I kinda agree that it can be hardcoded, because its logging, and will never be in released code (AFAIK), you could consider if you want to add some kconfig option for this though.

hakonfam commented on boot/zephyr/main.c:
```diff
@@ -175,6 +183,40 @@ static void do_boot(struct boot_rsp *rsp)
+
+     while (1) {
+             if (log_process(false) == false) {
+                    k_sleep(30);
```
Magic value `30` should have some define with a comment behind its rationale.

mbolivar commented on boot/zephyr/main.c:
```diff
@@ -175,6 +183,40 @@ static void do_boot(struct boot_rsp *rsp)
 
+#if defined(CONFIG_LOG) && !defined(CONFIG_LOG_IMMEDIATE) &&\
+    !defined(CONFIG_LOG_PROCESS_THREAD)
+#include <logging/log_ctrl.h>
```
Why is it necessary to do this twice instead of just once, outside of `CONFIG_SOC_FAMILY_NRF`?

nvlsianpu commented:
> Do you need to shut down this thread before jumping to the application? CC @nordic-krch

No it doesn't need that.

nvlsianpu commented:
@mbolivar @hakonfam DNM removed, passed zephyr sanitycheck locally, yours comments were taken into account.

mbolivar commented on boot/zephyr/main.c:
```diff
@@ -44,6 +44,28 @@ const struct boot_uart_funcs boot_funcs = {
+#else
+#include <logging/log_ctrl.h>
+
+#define BOOT_LOG_STACK_SIZE 768
```
I think @hakonfam 's comment about making this Kconfigurable (and the log processing interval too) is not addressed.

mbolivar replied:
> I think @hakonfam 's comment about making this Kconfigurable (and the log processing interval too) is not addressed.

Do you intend to address this one?

nvlsianpu replied:
addressed - I followed log internal thread stack size from zephyr codebase

mbolivar commented:
Thanks! Looking much more generic now.



> > Do you need to shut down this thread before jumping to the application? CC @nordic-krch
> 
> No it doesn't need that.

I have some doubts here. Can you explain your reasoning?

nordic-krch commented:
> Do you need to shut down this thread before jumping to the application?

I assume that bootloader simply jumps to the application, right? If so then i'm a bit afraid about what happens with system clock which is running (RTC) and will expire after jumping to the application. For application kernel that will be confusing and may lead to serious, hard to detect issues.

I don't know mcuboot details but imo jumping to the application from bootloader should always happen after reset (with clean state). There are some resources started (usb, clock, rtc, mpu?, etc.) and application is not expecting that.

nvlsianpu commented:
> Do you need to shut down this thread before jumping to the application? CC @nordic-krch

@mbolivar @nordic-krch I didn't observed any problem which causes any system failure.
What I'm aware is #529 issue, what you are afraid is the same problem.

I can kill the log processing thread - but this still not guaranties that RTOS release/clean all HW clocks and timers resources (I beet it doesn't).

@nordic-krch 
I was thinking about chain-load the application via reboot before jumping to the application might create Time of check, time of use issue. I would consider that execution flow on separate feature-discussion. Do you want to start the discussion via mcboot's feature issue creation?

@utzig How similar problem is solved for Mynewt

nordic-krch commented:
> I was thinking about chain-load the application via reboot before jumping to the application might create Time of check, time of use issue.

It's not about validation of the application. That should be performed before jump to application or omitted only on user responsibility, e.g. after special reset like wakeup from deep sleep. From my experience with the bootloader it nrf5 sdk, there should be no direct jump to the application once bootloader performed any actions other than the one needed to start application. That is image update or recovery. It's usually quite difficult to get back to clean state and reset is the safest and simplest way.

> I didn't observed any problem

That can be very random. Especially with active timeout.

nvlsianpu commented:
@mbolivar @nordic-krch In the most recent commit I added termination of the log processing thread.

mbolivar commented on boot/zephyr/main.c:
```diff
@@ -175,6 +200,58 @@ static void do_boot(struct boot_rsp *rsp)
+    boot_log_stop = true;
+
+    /* busy wait until log procesing thread expired*/
+    while (boot_log_stop) {
```
Are you sure this is enough? Might be worth using k_kthread_abort() afterwards:

https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#_CPPv414k_thread_abort7k_tid_t

(I just am being paranoid about cleaning up according to Zephyr API promises, rather than what happens to work today.)

nvlsianpu replied:
quote [thread termination behavior description](
https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#thread-termination):
> Once a thread is started it typically executes forever. However, a thread may synchronously end its execution by returning from its entry point function. This is known as termination.

In the next paragraph (about aborting):
> it is typically preferable to signal a thread to terminate itself gracefully, rather than aborting it.

So thread ends execution and that all needed. I might call abort but according to this, it is not  required.

mbolivar replied:
Right, but the current implementation looks racy. How do you know the thread is done, instead of just having set the variable and gotten scheduled out? Are all the threads cooperative?

mbolivar replied:
Looking at https://github.com/zephyrproject-rtos/zephyr/issues/21500, it seems like the common pattern right now is to use a semaphore given at the end of the exiting thread, since there is no `k_thread_join()` in Zephyr.

 From that issue:

> There's no good way to block until another thread exits.
> 
> Our test cases currently manage this in a few ways:
>
>    1. Create an equal priority co-op thread, or higher priority thread, and then yield() the CPU, or sleep for a while. Assuming the child thread never blocks and then exits, control returns after yield() completes...if we are in uniprocessor. In SMP the child could spawn on another CPU in which case after the yield() returns the child is probably still running. The workaround is the 1cpu variant of test case declarations, which spawn a busy-looping co-op thread on all other CPUs.
>    2. Synchronize by having the parent take a semaphore given by the child at the end of its entry function. Potentially race-prone if the child hasn't completely exited and the parent tries to re-use the thread object, resulting in a crash, but still safer than #1.

So I think on Nordic SoCs, the current implementation, which uses a higher priority thread and a flag global, is probably safe. But just for the sake of keeping to the common pattern, how about using approach 2. for now (give a semaphore at log thread exit), and add a comment to switch to using `k_thread_join()` as soon as that is available?

nvlsianpu replied:
^^ updated.
I could even use the same semaphore (but numeric one) to request thread termination - but I fell this as a overkill, so kept trigger via volatile variable,

mbolivar commented on boot/zephyr/main.c:
```diff
@@ -216,7 +299,7 @@ void main(void)
                             GPIO_DIR_IN | GPIO_PUD_PULL_UP
 #endif
-	    );
+           );
```
Unnecessary whitespace change

nvlsianpu replied:
I replaced Tab indentation. spaces indentation is mcuboot CS contention,

mbolivar commented on boot/zephyr/main.c:
```diff
@@ -175,6 +200,62 @@ static void do_boot(struct boot_rsp *rsp)
+{
+    boot_log_stop = true;
+
+    /* busy wait until log procesing thread expired
```
Technically this is not busy waiting, because calling k_sem_take should put the thread to sleep until the semaphore is available. 

```suggestion
    /* wait until log procesing thread expired
```
